cpp/include/cudf/detail/utilities/device_operators.cuh

```diff
@@ -55,6 +55,34 @@ CUDF_HOST_DEVICE inline auto max(LHS const& lhs, RHS const& rhs)
 {
   return std::max(lhs, rhs);
 }
+
+/**
+ * @brief Functor that casts another functor's result to a specified type.
+ *
+ * CUB 2.0.0 reductions require that the binary operator returns the same type
+ * as the initial value type, so we wrap binary operators with this when used
+ * by CUB.
+ */
+template <typename ResultType, typename F>
+struct cast_functor_fn {
```

Comment:
I'm open to other ideas on where this functor should go. I put it in `device_operators.cuh` because it wraps binary device operators used for reduction/scans. It's reused a handful of times in this PR so it ought to be put somewhere common. It could be in its own header, like `cudf/detail/utilities/cast_functor.cuh` if preferred.

Reply:
Based on how it is used currently in this PR, I would vote for placing this into its own header file.

Reply:
Done in a4b359f23ee0b6c9e23a455a902f87989d6692f5.